### PR TITLE
EventHubs tracing and metrics: catch up with OTel specification

### PR DESCRIPTION
**Pull request by lmolkova**
# Description

- Updates EventHubs to the latest otel semconv (yet to be released, so this PR will need to wait).
- Refactors how tracing and metrics are done to minimize their effect on calling code and maximize testability.
- Adds tons of non-mocking unit tests for tracing and metrics.

It addresses most of #37445 part for EH (all but feature flag) and significant portion of #38351

(PS: I think I finally found the right pattern to trace reactor publishers 😅)

### CODE REVIEW

File: eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml [hunk 1]
```diff
@@ -93,9 +93,8 @@
   <suppress checks="IllegalImport" files="com.azure.messaging.servicebus.IntegrationTestBase.java"/>
   <suppress checks="IllegalImport" files="com.azure.messaging.servicebus.TracingIntegrationTests.java"/>
   <suppress checks="IllegalImport" files="com.azure.messaging.servicebus.LoggingSpanProcessor.java"/>
-  <suppress checks="IllegalImport" files="com.azure.messaging.eventhubs.TracingIntegrationTests.java"/>
-  <suppress checks="IllegalImport" files="com.azure.messaging.eventhubs.PublishEventsTracingWithCustomContextSample.java"/>
-  <suppress checks="IllegalImport" files="com.azure.messaging.eventhubs.EventProcessorWithTracingClientSample.java"/>
+  <suppress checks="IllegalImport" files="com.azure.messaging.eventhubs..*Sample.*.java"/>
```
Comment (conniey) [inline eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml:96]: Since it's using regex, the `.` is "match anything" rather than the period we want to match against. I mean, it doesn't change anything wrt to being able to match it. :) But I thought it was funny.

File: sdk/core/azure-core-tracing-opentelemetry/src/test/java/com/azure/core/tracing/opentelemetry/OpenTelemetryHttpPolicyTests.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ public void openTelemetryHttpPolicyTest() {
         request.setHeader(HttpHeaderName.USER_AGENT, "user-agent");
 
         try (Scope scope = parentSpan.makeCurrent()) {
-            createHttpPipeline(azTracer).send(request, tracingContext).block().close();
+            createHttpPipeline(azTracer).send(request, tracingContext).block().getBodyAsBinaryData();
```
Comment (conniey) [inline sdk/core/azure-core-tracing-opentelemetry/src/test/java/com/azure/core/tracing/opentelemetry/OpenTelemetryHttpPolicyTests.java:115]: I'm curious why the change from close to getBodyAsBinaryData?

Comment (lmolkova) [reply sdk/core/azure-core-tracing-opentelemetry/src/test/java/com/azure/core/tracing/opentelemetry/OpenTelemetryHttpPolicyTests.java:115]: to test that we end span when we end reading the body and not just when the response is closed  - apparently we're not great at closing response, but very good at reading body to the end on storage.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/InstrumentationPolicy.java [hunk 3]
```diff
@@ -199,56 +199,65 @@ public HttpHeaders getHeaders() {
 
         @Override
         public Flux<ByteBuffer> getBody() {
-            return response.getBody().doOnError(e -> exception = e).doOnCancel(() -> errorType = CANCELLED_ERROR_TYPE);
+            return Flux.using(() -> span,
+                s -> response.getBody()
+                    .doOnError(e -> tracer.end(null, e, s))
+                    .doOnCancel(() -> tracer.end(CANCELLED_ERROR_TYPE, null, s)),
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/InstrumentationPolicy.java]: it's ok, but fixed anyway in #39381

**Comment by lmolkova** [comment]
dear bot, please don't close it.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsMetricsProvider.java [hunk 4]
```diff
@@ -3,94 +3,208 @@
...
+import static com.azure.messaging.eventhubs.implementation.instrumentation.InstrumentationUtils.MESSAGING_SYSTEM_VALUE;
+import static com.azure.messaging.eventhubs.implementation.instrumentation.InstrumentationUtils.SERVER_ADDRESS;
+import static com.azure.messaging.eventhubs.implementation.instrumentation.InstrumentationUtils.getDurationInSeconds;
+import static com.azure.messaging.eventhubs.implementation.instrumentation.InstrumentationUtils.getOperationType;
+import static com.azure.messaging.eventhubs.implementation.instrumentation.OperationName.CHECKPOINT;
+import static com.azure.messaging.eventhubs.implementation.instrumentation.OperationName.PROCESS;
+import static com.azure.messaging.eventhubs.implementation.instrumentation.OperationName.RECEIVE;
+import static com.azure.messaging.eventhubs.implementation.instrumentation.OperationName.SEND;
 
 public class EventHubsMetricsProvider {
```
Comment (anuchandy) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsMetricsProvider.java:40]: Does it make sense to mark all our tracing implementation classes as `final` in case we enable revApi for impl types later, as discussed yesterday?

**Review by anuchandy — APPROVED** [review]
thank you!

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationUtils.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            case SEND:
+                return "publish";
+            case RECEIVE:
+                return "receive";
+            case CHECKPOINT:
+                return "settle";
+            case PROCESS:
+                return "process";
+            default:
+                return null;
```
Comment (anuchandy) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationUtils.java:97]: Is it fine to use `null` for OperationName.GET_EVENT_HUB_PROPERTIES  and OperationName.GET_PARTITION_PROPERTIES values?

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationScope.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    public String getErrorType() {
+        if (errorType == null) {
+            errorType = InstrumentationUtils.getErrorType(error);
+        }
+        return errorType;
+    }
+
+    @Override
+    public void close() {
+        if (!closed) {
```
Comment (anuchandy) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationScope.java:94]: since we've this guard check, do we know if there a potential for `close()` to be called from two different threads? if so, wondering `closed` need to be `volatile` or we're safe

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,8 @@
 
 ### Breaking Changes
 
+- Experimental checkpointing metrics are no longer reported by this package. They've been moved to `azure-messaging-eventhubs` package.
```
Comment (lmolkova) [inline sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/CHANGELOG.md:9]: TODO link to PR

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationScope.java (hunk 6 above)
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationScope.java:94]: it should not be executed concurrently, but let's stay on the safe side and assume it could be. So changed to volatile. Thank you!

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessorClient.java [hunk 8]
```diff
@@ -292,6 +298,15 @@ public synchronized void stop() {
...
     }
 
+    synchronized boolean stop(long timeout, TimeUnit timeUnit) {
+        stop();
+        try {
+            return scheduler.get().awaitTermination(timeout, timeUnit);
+        } catch (InterruptedException e) {
+            return false;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessorClient.java]: yep, removed now that #41878 is merged.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsMetricsProvider.java [hunk 9]
```diff
@@ -3,94 +3,208 @@
...
+        }
+
+        setOperation(attributes, operationName);
+        attributes.put(ERROR_TYPE, errorType);
+        return meter.createAttributes(attributes);
     }
 
-    class AttributeCache {
+    private Map<String, Object> getCommonAttributes(String namespace, String entityName, String consumerGroup) {
+        Map<String, Object> commonAttributesMap = new HashMap<>(3);
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsMetricsProvider.java]: good point!

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationUtils.java (hunk 5 above)
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationUtils.java:100]: they're intentionally null - added the following code comment to explain

> other operations are not documented in the otel semantic conventions
> they are not 'core' messaging operations and therefor don't have type
> we'll still report operation name attribute for them.

**Comment by ikechukwu1987432** [comment]
automatic support

**Comment by ikechukwu1987432** [comment]
automatic setup

**Comment by ikechukwu1987432** [comment]
@ikechukwu1987432

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/CHANGELOG.md (hunk 7 above)
Comment (ikechukwu1987432) [reply sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/CHANGELOG.md:9]: -isinweke
